exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -376,100 +376,63 @@ func (r *MachinePoolReconciler) reconcileMachines(ctx context.Context, mp *expv1
 	}
 
 	return nil
 }
 
-// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
-func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) ([]clusterv1.Machine, error) {
+// createOrUpdateMachines creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createOrUpdateMachines(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
 	log := ctrl.LoggerFrom(ctx)
 
 	// Construct a set of names of infraMachines that already have a Machine.
-	infraRefNames := sets.Set[string]{}
+	infraMachineToMachine := map[string]clusterv1.Machine{}
 	for _, machine := range machines {
 		infraRef := machine.Spec.InfrastructureRef
-		infraRefNames.Insert(infraRef.Name)
+		infraMachineToMachine[infraRef.Name] = machine
 	}
 
 	createdMachines := []clusterv1.Machine{}
 	var errs []error
 	for i := range infraMachines {
 		infraMachine := &infraMachines[i]
-		// If infraMachine already has a Machine, skip it.
-		if infraRefNames.Has(infraMachine.GetName()) {
-			log.V(4).Info("Machine already exists for infraMachine", "infraMachine", klog.KObj(infraMachine))
-			continue
-		}
-		// Otherwise create a new Machine for the infraMachine.
-		log.Info("Creating new Machine for infraMachine", infraMachine.GroupVersionKind().Kind, klog.KObj(infraMachine))
-		machine := getNewMachine(mp, infraMachine)
-		if err := r.Client.Create(ctx, machine); err != nil {
-			errs = append(errs, errors.Wrapf(err, "failed to create new Machine for infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace()))
-			continue
+		// If infraMachine already has a Machine, update it if needed.
+		if existingMachine, ok := infraMachineToMachine[infraMachine.GetName()]; ok {
+			log.V(2).Info("Patching existing Machine for infraMachine", "infraMachine", klog.KObj(infraMachine), "machine", klog.KObj(&existingMachine))
+
+			desiredMachine := computeDesiredMachine(mp, infraMachine, &existingMachine)
+			if err := ssa.Patch(ctx, r.Client, MachinePoolControllerName, desiredMachine, ssa.WithCachingProxy{Cache: r.ssaCache, Original: &existingMachine}); err != nil {
+				log.Error(err, "failed to update Machine", "Machine", klog.KObj(desiredMachine))
+				errs = append(errs, errors.Wrapf(err, "failed to update Machine %q", klog.KObj(desiredMachine)))
+			}
+		} else {
+			// Otherwise create a new Machine for the infraMachine.
+			log.Info("Creating new Machine for infraMachine", "infraMachine", klog.KObj(infraMachine))
+			machine := computeDesiredMachine(mp, infraMachine, nil)
+
+			if err := r.Client.Create(ctx, machine); err != nil {
```

Comment:
@CecileRobertMichon @Jont828 

I would recommend using the following here:
```go
if err := ssa.Patch(ctx, r.Client, MachinePoolControllerName, machine); err != nil {
```

Otherwise you end up with co-ownership of fields between "manager" and "machinepool-controller" like here: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1725100075312484352/artifacts/clusters/bootstrap/resources/quick-start-jwp00u/Machine/quick-start-loyh5w-mp-0-9svll-hk7jp.yaml

The consequence is that it's impossible to unset fields via "machinepool-controller"

Reply:
I'll take a look as part of https://github.com/kubernetes-sigs/cluster-api/pull/9725/files

Reply:
Thx!